### PR TITLE
NMS-12079: eliminate create.sql in favor of Liquibase for integration tests

### PR DESCRIPTION
**Pull request by deejgregor**
### All Contributors

* [x] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [x] Have you made a comment in that issue which points back to this PR?
* [x] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [x] If this is a new or updated feature, is there documentation for the new behavior?
* [x] If this is new code, are there unit and/or integration tests?
* ~~If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?~~

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

Once the reviewer(s) accept the PR and the branch passes continuous integration in Bamboo, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA: https://issues.opennms.org/browse/NMS-12079
* Bamboo: https://bamboo.opennms.org/browse/OPENNMS-OPENNMS85
* CircleCI: https://circleci.com/gh/OpenNMS/opennms/tree/jira%2FHZN-1111
* Wiki article on Integration tests: https://wiki.opennms.org/wiki/Integration_tests

### CODE REVIEW

File: core/test-api/db/src/test/java/org/opennms/core/test/db/IPLikeCoverageIT.java [hunk 1]
```diff
@@ -28,21 +28,41 @@
...
+import java.sql.ResultSet;
 
-public class IPLikeCoverageIT extends InstallerDbITCase {
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.opennms.core.utils.DBUtils;
+
+@Ignore("ignoring for now to see if we can get things passing")
```
Comment (RangerRick) [inline core/test-api/db/src/test/java/org/opennms/core/test/db/IPLikeCoverageIT.java]: should this be un-ignored?

File: core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java [hunk 2]
```diff
@@ -73,23 +80,29 @@
...
+    public void setTemporaryDatabase(TemporaryDatabase database) {
+        m_database = database;
+    }
+
     @Before
     public void setUp() throws Exception {
         System.setProperty("org.apache.logging.log4j.simplelog.StatusLogger.level", "INFO");
         MockLogAppender.setupLogging();
+
+        TemporaryDatabasePostgreSQL temp = new TemporaryDatabasePostgreSQL();
```
Comment (RangerRick) [inline core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java]: AFAICT this does nothing, or are you doing it just for initialization?  If so you don't need the variable.

File: core/test-api/db/src/test/java/org/opennms/core/test/db/TemporaryDatabasePostgreSQLIT.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import java.io.IOException;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.opennms.core.schema.Migrator;
+import org.opennms.core.test.MockLogAppender;
+import org.springframework.context.support.GenericApplicationContext;
+import org.springframework.context.support.StaticApplicationContext;
+
+public class TemporaryDatabasePostgreSQLIT {
```
Comment (RangerRick) [inline core/test-api/db/src/test/java/org/opennms/core/test/db/TemporaryDatabasePostgreSQLIT.java]: This IT does a lot of "run things" but not very much "check it actually made and deleted the database as we expect".  You should at least have an assert in each `@Test`.  :)

File: features/ticketing/daemon/src/test/java/org/opennms/netmgt/ticketd/DefaultTicketerServiceLayerIT.java [hunk 4]
```diff
@@ -200,6 +201,7 @@ public void testTicketsForSituations() throws PluginException {
         assertFalse(relatedAlarms.isEmpty());
         assertEquals(relatedAlarms.get(0).getNodeId(), testNode.getId());
 
+        m_alarmDao.delete(situation); // XXX this is working around a bug with deleting an alarm that is part of a situation
```
Comment (RangerRick) [inline features/ticketing/daemon/src/test/java/org/opennms/netmgt/ticketd/DefaultTicketerServiceLayerIT.java]: Do you have more info on this bug?  Is there an open issue about it?

File: features/ticketing/daemon/src/test/java/org/opennms/netmgt/ticketd/DefaultTicketerServiceLayerIT.java [hunk 5]
```diff
@@ -109,6 +109,7 @@ public void setUp() {
 
     @After
     public void tearDown() {
+
```
Comment (mvrueden) [inline features/ticketing/daemon/src/test/java/org/opennms/netmgt/ticketd/DefaultTicketerServiceLayerIT.java]: Unnecessary whitespace.

File: core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java [hunk 6]
```diff
@@ -181,29 +200,41 @@ public void testMultipleChangelogs() throws Exception {
         assertEquals("test-api.schema.b", ids.get(1).getId());
     }
 
+    public static Predicate<Resource> createTestApiLiquibaseChangelogFilter() {
```
Comment (fooker) [inline core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java]: This can be a final static constant

File: core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java [hunk 7]
```diff
@@ -181,29 +200,41 @@ public void testMultipleChangelogs() throws Exception {
...
+        return r -> {
+            try {
+                URI uri = r.getURI();
+                return (uri.getScheme().equals("file") && uri.toString().contains("test-api/schema")) ||
+                        (uri.getScheme().equals("jar") && uri.toString().contains("test-api.schema"));
+            } catch (IOException e) {
+                return false;
```
Comment (fooker) [inline core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java]: Maybe this should just propagate?

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.pathoutage/src/test/java/org/opennms/features/topology/plugins/topo/pathoutage/PathOutageStatusProviderIT.java [hunk 8]
```diff
@@ -323,21 +335,14 @@ private OnmsOutage createOutage(String uie, String ipaddress, OnmsNode node, Onm
 	 * @return Resulting service
 	 */
 	private OnmsMonitoredService createService(OnmsNode node) {
-		OnmsMonitoredService monitoredService = new OnmsMonitoredService();
-		OnmsServiceType serviceType = createType("someType");
+		OnmsServiceType serviceType = new OnmsServiceType();
+		serviceType.setName("someType" + ++s_serviceTypeCounter);
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.pathoutage/src/test/java/org/opennms/features/topology/plugins/topo/pathoutage/PathOutageStatusProviderIT.java]: Can we not do this in one line?

File: core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java [hunk 9]
```diff
@@ -352,24 +382,45 @@ public String toString() {
...
+    private static Collection<String> getClassLoaderUrls(ApplicationContext context) {
+        List<String> urls = new LinkedList<String>();
+        for (ApplicationContext c = context; c != null; c = c.getParent()) {
+            for (ClassLoader l = c.getClassLoader(); l != null; l = l.getParent()) {
+                if (l instanceof URLClassLoader) {
+                    for (URL u : ((URLClassLoader)l).getURLs()) {
+                        urls.add(u.toString());
+                    }
+                } else {
+                    System.err.println("*** Couldn't get classloader URLs for classloader " + l + ", not an instance of URLClassLoader");
```
Comment (fooker) [inline core/test-api/db/src/test/java/org/opennms/core/test/db/MigratorIT.java]: If this is an error, we should throw something

File: core/web-assets/src/main/java/org/opennms/web/assets/impl/AssetLocatorImpl.java [hunk 10]
```diff
@@ -144,7 +144,7 @@ public long lastModified() {
 
     @Override
     public Optional<InputStream> open(final String assetName, final String type, final boolean minified) throws IOException {
-        return withLogPrefix(() -> {
+        return this.<Optional<InputStream>>withLogPrefix(() -> {
```
Comment (fooker) [inline core/web-assets/src/main/java/org/opennms/web/assets/impl/AssetLocatorImpl.java]: Does this really needs to be fully qualified?

**Comment by j-white** [comment]
I ran into the following error will trying to run the HttpCollectorIT locally: https://gist.github.com/j-white/8fb44587d33189e70131f993b7ac5a3a

File: core/test-api/db/src/main/java/org/opennms/core/test/db/MockDatabase.java [hunk 11]
```diff
@@ -84,7 +84,7 @@ public MockDatabase(String name, boolean createNow) throws Exception {
 
         setClassName(MockDatabase.class.getName());
         setMethodName("MockDatabase constructor");
-        setTestDetails("I do not know who called me.... which is sad. Will you be my friend?");
+        setTestDetails("I do not know who called me.... which is sad. Will you be my friend and help me fix this?");
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply core/test-api/db/src/main/java/org/opennms/core/test/db/MockDatabase.java]: @mvrueden What do you mean by open an issue here?  You mean add to the text to request opening an issue?

[the remark this answers is not in the document]
Comment (mvrueden) [reply core/test-api/db/src/main/java/org/opennms/core/test/db/MockDatabase.java]: Nah, open an issue which actually fixes/addresses the mentioned problem :)

**Comment by j-white** [comment]
Closing PR since there's still some work left to do here and no progress has been made recently.

File: features/ticketing/daemon/src/test/java/org/opennms/netmgt/ticketd/DefaultTicketerServiceLayerIT.java (hunk 4 above)
Comment (RangerRick) [reply features/ticketing/daemon/src/test/java/org/opennms/netmgt/ticketd/DefaultTicketerServiceLayerIT.java]: turns out this is a cascade issue with the alarm association object, I _think_ I have a fix

File: features/flows/classification/engine/impl/src/test/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngineIT.java [hunk 12]
```diff
@@ -151,7 +150,7 @@ public void verifyRuleFilter() {
         assertThat(classificationEngine.classify(classificationRequest), is("rule4"));
 
         // Update no rule matches
-        classificationRequest.setDstPort(443);
+        classificationRequest.setDstPort(0);
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply features/flows/classification/engine/impl/src/test/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngineIT.java:153]: I think this was changed because now we're getting the full ruleset from liquibase and there is a match for 443 (`https`)

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.pathoutage/src/test/java/org/opennms/features/topology/plugins/topo/pathoutage/PathOutageStatusProviderIT.java (hunk 8 above)
Comment (RangerRick) [reply features/topology-map/plugins/org.opennms.features.topology.plugins.topo.pathoutage/src/test/java/org/opennms/features/topology/plugins/topo/pathoutage/PathOutageStatusProviderIT.java]: yup

File: core/test-api/db/src/test/java/org/opennms/core/test/db/IPLikeCoverageIT.java [hunk 13]
```diff
@@ -119,14 +139,35 @@ public void testIplikeCoverage() throws Exception {
...
+            util.watch(st);
+            st.setString(1, value);
+            st.setString(2, rule);
+
+            st.execute();
+
+            rs = st.getResultSet();
+            util.watch(rs);
+            final boolean result = rs.getBoolean(1);
+            if (expected) {
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply core/test-api/db/src/test/java/org/opennms/core/test/db/IPLikeCoverageIT.java]: yup

File: core/test-api/db/src/main/java/org/opennms/core/test/db/MockDatabase.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (RangerRick) [reply core/test-api/db/src/main/java/org/opennms/core/test/db/MockDatabase.java]: Not sure how easy it is to figure out what's going on, but I've at least added a stack trace to the output in case we see it more.

**Comment by RangerRick** [comment]
Last failure on smoke tests was just because I tried setting them to run full tests, but they weren't parallelized, so it timed out.  Went ahead and put everything back.  @deejgregor what do you think?  Should we merge this?  You actually got it basically working last year, there were just a few more tweaks to do. 👍
